### PR TITLE
feat: sum tvls of all protocols per chain

### PR DESCRIPTION
**Pull request by Addis-Belete**
**NOTE**

#### Please enable "Allow edits by maintainers" while putting up the PR.

---

> - If you would like to add a `volume` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
3. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
4. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
5. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
6. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---

### CODE REVIEW

**Comment by llamatester** [comment]
Error while running adapter at projects/clearpool: 
        
 
  
    
    
    [object Object]

**Comment by llamatester** [comment]
Error while running adapter at projects/clearpool: 
        
 
  
    
    
    [object Object]

**Comment by Addis-Belete** [comment]
Screenshot 2024-05-15 at 1 53 19 PM

**Comment by llamatester** [comment]
The adapter at projects/clearpool exports TVL: 
        
 
 
    ethereum-staking          20.38 M
    staking                   20.38 M
    borrowed                  17.25 M
    optimism-borrowed         10.61 M
    ethereum-borrowed         5.55 M
    optimism                  2.13 M
    ethereum                  1.48 M
    avax-borrowed             640.32 k
    mantle-borrowed           280.90 k
    polygon_zkevm-borrowed    174.58 k
    mantle                    47.61 k
    polygon                   41.15 k
    polygon_zkevm             32.18 k
    avax                      45
    polygon-borrowed          0
    arbitrum                  0
    arbitrum-borrowed         0
    flare                     0
    flare-borrowed            0
    base-borrowed             0
    base                      0
    
    total                    3.72 M

File: projects/clearpool/index.js [hunk 1]
```diff
@@ -125,10 +126,32 @@ Object.keys(config).forEach((protocol) => {
     const borrowed = async (api) => {
       const { pools, tokens } = await _getLogs(api);
       const bals = await api.multiCall({ abi: borrowFn, calls: pools });
-      api.addTokens(tokens, bals);
+     return api.addTokens(tokens, bals);
     };
+    if (!module.exports[chain]) {
```
Comment (georgeciubotaru) [inline projects/clearpool/index.js]: @Addis-Belete The code is good, but we need a more readable version to be easily understood.

Comment (georgeciubotaru) [reply projects/clearpool/index.js]: let's use the `api` methods that they provide maybe we can somehow benefit from them

**Comment by llamatester** [comment]
Error while running adapter at projects/clearpool: 
        
 
  
    
    
    [object Object]

**Comment by llamatester** [comment]
The adapter at projects/clearpool exports TVL: 
        
 
 
    ethereum-staking          20.20 M
    staking                   20.20 M
    borrowed                  17.41 M
    optimism-borrowed         10.76 M
    ethereum-borrowed         5.56 M
    optimism                  2.16 M
    ethereum                  1.48 M
    avax-borrowed             640.50 k
    mantle-borrowed           281.75 k
    polygon_zkevm-borrowed    174.77 k
    mantle                    47.75 k
    polygon                   41.16 k
    polygon_zkevm             32.22 k
    avax                      45
    polygon-borrowed          0
    arbitrum                  0
    arbitrum-borrowed         0
    flare                     0
    flare-borrowed            0
    base-borrowed             0
    base                      0
    
    total                    3.76 M

**Comment by Addis-Belete** [comment]
Log: getting undefined while calculating totalBorrowed

**Comment by georgeciubotaru** [comment]
@Addis-Belete Did you resolve the issue?

**Comment by Addis-Belete** [comment]
@georgeciubotaru I am implementing in another way. will push soon

**Comment by llamatester** [comment]
Error while running adapter at projects/clearpool: 
        
 
  
    
    
    [object Object]

**Comment by llamatester** [comment]
Error while running adapter at projects/clearpool: 
        
 
  
    
    
    [object Object]

**Comment by Addis-Belete** [comment]
Screenshot 2024-05-15 at 6 51 06 PM

**Comment by Addis-Belete** [comment]
@georgeciubotaru Review now

**Comment by Addis-Belete** [comment]
Log: used additional 2hr to complete this PR.

File: projects/clearpool/index.js [hunk 2]
```diff
@@ -92,44 +104,106 @@ const getEventAndABI = (protocol) => {
   }
   return { borrowFn, abi };
 };
+let adapters = {};
```
Comment (georgeciubotaru) [inline projects/clearpool/index.js]: 1. get rid of `adapters` as now you can assign directly to the `module.exports`

2. `prepareProtocolsPerChain` is called on every chain (line 133), you can optimize by passing chain as param and prepare only for that chain


Let's quickly do these and integrate the other protocol in the other PR.

Good job tho!! 👍

**Comment by llamatester** [comment]
Error while running adapter at projects/clearpool: 
        
 
  
    
    
    [object Object]

File: projects/clearpool/index.js [hunk 3]
```diff
@@ -93,43 +105,97 @@ const getEventAndABI = (protocol) => {
...
+      const { abi, borrowFn } = getEventAndABI(protocol);
+      const data = {
+        fromBlock,
+        factory,
+        abi,
+        borrowFn,
+        protocol,
+      };
+        contracts.push(data);
+    });
```
Comment (georgeciubotaru) [inline projects/clearpool/index.js]: formatting here

**Comment by llamatester** [comment]
The adapter at projects/clearpool exports TVL: 
        
 
 
    ethereum-staking          22.57 M
    staking                   22.57 M
    borrowed                  17.51 M
    optimism-borrowed         10.85 M
    ethereum-borrowed         5.57 M
    optimism                  2.40 M
    ethereum                  1.46 M
    avax-borrowed             642.07 k
    mantle-borrowed           282.12 k
    polygon_zkevm-borrowed    174.83 k
    mantle                    77.41 k
    polygon                   41.25 k
    polygon_zkevm             32.22 k
    avax                      45
    polygon-borrowed          0
    arbitrum                  0
    arbitrum-borrowed         0
    flare                     0
    flare-borrowed            0
    base-borrowed             0
    base                      0
    
    total                    4.00 M

File: projects/clearpool/index.js [hunk 4]
```diff
@@ -93,43 +105,97 @@ const getEventAndABI = (protocol) => {
...
+      const { fromBlock, factory } = config[chain][protocol];
+      const { abi, borrowFn } = getEventAndABI(protocol);
+      const data = {
+        fromBlock,
+        factory,
+        abi,
+        borrowFn,
+        protocol,
+      };
+        contracts.push(data);
```
Comment (georgeciubotaru) [inline projects/clearpool/index.js]:
```suggestion
      const data = {
        fromBlock,
        factory,
        abi,
        borrowFn,
        protocol,
      };
        contracts.push({
         fromBlock
         factory
         abi
         borrowFn
         protocol
        });
```

**Comment by llamatester** [comment]
Error while running adapter at projects/clearpool: 
        
 
  
    
    
    [object Object]

**Comment by llamatester** [comment]
The adapter at projects/clearpool exports TVL: 
        
 
 
    ethereum-staking          22.57 M
    staking                   22.57 M
    borrowed                  17.51 M
    optimism-borrowed         10.85 M
    ethereum-borrowed         5.57 M
    optimism                  2.40 M
    ethereum                  1.46 M
    avax-borrowed             642.07 k
    mantle-borrowed           282.13 k
    polygon_zkevm-borrowed    174.83 k
    mantle                    77.41 k
    polygon                   41.25 k
    polygon_zkevm             32.22 k
    avax                      45
    polygon-borrowed          0
    arbitrum-borrowed         0
    flare                     0
    flare-borrowed            0
    arbitrum                  0
    base-borrowed             0
    base                      0
    
    total                    4.00 M

**Comment by llamatester** [comment]
The adapter at projects/clearpool exports TVL: 
        
 
 
    ethereum-staking          22.57 M
    staking                   22.57 M
    borrowed                  17.46 M
    optimism-borrowed         10.80 M
    ethereum-borrowed         5.56 M
    optimism                  2.39 M
    ethereum                  1.46 M
    avax-borrowed             641.43 k
    mantle-borrowed           282.13 k
    polygon_zkevm-borrowed    174.83 k
    mantle                    77.41 k
    polygon                   41.20 k
    polygon_zkevm             32.22 k
    avax                      45
    polygon-borrowed          0
    arbitrum-borrowed         0
    arbitrum                  0
    flare-borrowed            0
    flare                     0
    base                      0
    base-borrowed             0
    
    total                    3.99 M

**Review by georgeciubotaru — APPROVED** [review]

**Pull request merged**